File: C#/0015. 3Sum.cs
public class Solution {
    public IList<IList<int>> ThreeSum(int[] nums) {
        IList<IList<int>> threeSumList=new List<IList<int>>();
        Array.Sort(nums);
        int cur=0;
        while(cur<nums.Length-1){
            IList<IList<int>> rep=new List<IList<int>>();
            rep=TwoSum(nums,cur);
            foreach(var list in rep){
                threeSumList.Add(list);
            }
            int target=nums[cur];
            cur++;
            while(cur<nums.Length && nums[cur]==target){
                cur++;
            }
        }
        return threeSumList;
    }
    public IList<IList<int>> TwoSum(int[] nums,int cur){
        IList<IList<int>> rep=new List<IList<int>>();
        int left=cur+1;
        int right=nums.Length-1;
        while(left<right){
            if(nums[left]+nums[right]+nums[cur]>0){
                int curr=nums[right];
                while(right>left && nums[right]==curr){
                    right--;
                }
            }
            else if (nums[left]+nums[right]+nums[cur]<0){
                int curl=nums[left];
                while(left<right && nums[left]==curl){
                    left++;
                }
            }
            else{        
                rep.Add(new List<int>{nums[cur],nums[left],nums[right]});
                int curl=nums[left];
                while(left<right && nums[left]==curl){
                    left++;
                }
            }
        }
        return rep;
    } 
}

File: 20190519python/0015. 3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        rep=[]
        
        k=0
        while k<len(nums):
            tmp=self.two_sum(nums[k+1:],-nums[k])
            for cur in tmp:
                rep.append([nums[k]]+cur)
            j=k
            while j<len(nums) and nums[j]==nums[k]:
                j+=1
            k=j
        return rep
    
    def two_sum(self,nums,target):
        rep=[]
        left,right=0,len(nums)-1
        while left<right:
            if nums[left]+nums[right]==target:
                rep.append([nums[left],nums[right]])
                k=left
                while k<right and nums[k]==nums[left]:
                    k+=1
                left=k
            elif nums[left]+nums[right]<target:
                left+=1
            else:
                right-=1
        return rep
